### PR TITLE
UI: Move fetching secret data to child route

### PR DESCRIPTION
**Pull request by hellobontempo**
### Description
Moves the request for secret data from the parent `secret.js` route to the child `secret.details` route

### TODO only if you're a HashiCorp employee
- [ ] **Backport Labels:** If this PR is in the ENT repo and needs to be backported, backport  
  to N, N-1, and N-2, using the `backport/ent/x.x.x+ent` labels. If this PR is in the CE repo, you should only backport to N, using the `backport/x.x.x` label, not the enterprise labels.
    - [ ] If this fixes a critical security vulnerability or [severity 1](https://www.hashicorp.com/customer-success/enterprise-support) bug, it will also need to be backported to the current [LTS versions](https://developer.hashicorp.com/vault/docs/enterprise/lts#why-is-there-a-risk-to-updating-to-a-non-lts-vault-enterprise-version) of Vault. To ensure this, use **all** available enterprise labels.
- [ ] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [ ] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [ ] **RFC:** If this change has an associated RFC, please link it in the description.
- [ ] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

### CODE REVIEW

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js
Comment (hellobontempo) [inline ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js:1450]: want to call out this change. I'm pretty sure that because the request to `/data/` has moved down we are prompted with another control group error here (after clicking the secret tab) and need to grant access. Want a gut check here that this makes sense!

File: ui/app/adapters/kv/metadata.js
Comment (hellobontempo) [inline ui/app/adapters/kv/metadata.js:81]: I do not understand how the delete tests were passing before 🤯 🤯 🤯  For the secret-creator persona there was no model (because the user can delete but not read) Ember magic??

File: ui/lib/kv/addon/components/kv-delete-modal.js
Comment (hellobontempo) [inline ui/lib/kv/addon/components/kv-delete-modal.js:23]: Looking at the code, this didn't seem true. It appeared that the metadata model was only required for `delete` which renders different delete options, please check me on this!

File: ui/lib/kv/addon/components/page/secret/metadata/details.js
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/secret/metadata/details.js:38]: This previously always returned a model so we could call `destroyRecord`, however, now we only fetch the secret model as a backup for reading `custom_metadata` so we don't want to rely on it for calling the delete. I opted to make a custom method instead since we only need the backend and path

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [inline ui/lib/kv/addon/routes/secret/metadata.js:10]: calling this service in the route file has been a newer pattern I've adopted during this subkey/patch work. Tests were failing because the model capability checks were not consistently returning what we expected in the test suite.

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [inline ui/lib/kv/addon/routes/secret/metadata.js:23]: This is new, so i'm not sure how folks will feel about it. It could be generalized and a method added to the capabilities service but I wanted to stew on it some more 🍲

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [inline ui/lib/kv/addon/routes/secret/metadata.js:19]: empty objects are truthy and we need falsy checks for the model hook

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [inline ui/lib/kv/addon/routes/secret/metadata.js:23]: the typical way of grabbing this info has been via the parentModel so updated for consistency

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [inline ui/lib/kv/addon/routes/secret/metadata.js:9]: My brain struggled with how to refactor this since it previously relied a _lot_ on the `kv/data` model. Open to other ways of handling this! I _think_ what I landed on was the least invasive, but a fresh perspective would be helpful!

File: ui/lib/kv/addon/components/page/list.js
Comment (hellobontempo) [inline ui/lib/kv/addon/components/page/list.js:90]: I reverted this because this test was failing and I felt like it made more sense to navigate straight to details in this instance (instead of checking whether the secret is real or not from the overview route)
https://github.com/hashicorp/vault/blob/7c2050a68f656cc1441c00b0b1dae9c5f0281dfe/ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-edge-cases-test.js#L175-L193

File: ui/lib/kv/addon/components/kv-delete-modal.js
Comment (Monkeychip) [reply ui/lib/kv/addon/components/kv-delete-modal.js:23]: I say this with little brain power at the moment. I believe the metadata-delete from the metadata model is used on the list view or the metadata tab view. So I believe you're correct. Probably a carry over from when we went back and forth if we were adding permanently delete to the modal, which we opted not to do.

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (Monkeychip) [reply ui/lib/kv/addon/routes/secret/metadata.js:23]: is the parent model always loaded/updated regardless of how you navigate to this route? ex: a `visit('this url')` in a test that goes directly to the route and doesn't touch the parent route.

File: ui/app/adapters/kv/metadata.js
Comment (Monkeychip) [reply ui/app/adapters/kv/metadata.js:81]: NIT: there is another API endpoint that is used to technically delete all versions — granted you have to pass an  an array of all versions. But the naming here threw me. For me `deleteMetadata` makes the most sense as it's the delete capability on the metadata endpoint. I know in the docs they say "Delete metadata and all versions" but we do have designs out there as a way to delete all versions using the ["Delete secret versions"](https://developer.hashicorp.com/vault/api-docs/secret/kv/kv-v2#delete-secret-versions) endpoint.

File: ui/lib/kv/addon/templates/secret/metadata/index.hbs
Comment (Monkeychip) [inline ui/lib/kv/addon/templates/secret/metadata/index.hbs:10]: That sneaky API and overlap. The `or` at first here didn't make sense to me because if the attr is `canReadCustomMetadata` then I would think we should only check read on metadata. Then I remembered that there's overlap that you set something on metadata and can read it on data. Maybe a comment here or in the component doc mentioning the overlap?

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (Monkeychip) [reply ui/lib/kv/addon/routes/secret/metadata.js:9]: Prior to the move-kv-to-ember-engine work, I had actually long ago added a capabilities check directly inside a kv component. KV is a unique case with coupled workflow and api endpoints complicating permission checks. All that to say, 1. I think calling the capability service here and not using the one from the model makes sense. 2. your structure of fetchCapabilities makes sense; kind of a fresh take on our more long handed approach we use now in the models. Nice work.

**Review by Monkeychip — COMMENTED** [review]
Nice work! Nothing blocking, but holding off an approval as I think Chelsea should also review; she's much more qualified to answer the control group questions.

File: ui/lib/kv/addon/components/page/list.js
Comment (hashishaw) [inline ui/lib/kv/addon/components/page/list.js:90]: Since this is a replacement for navigating in the list view, I disagree that we should go directly to details

File: ui/lib/kv/addon/components/page/secret/metadata/details.hbs
Comment (hashishaw) [inline ui/lib/kv/addon/components/page/secret/metadata/details.hbs:51]: Could we instead check for `this.customMetadata`? Since the read already happened I'm not sure why we need this capability

File: ui/lib/kv/addon/routes/secret/details.js
Comment (hashishaw) [inline ui/lib/kv/addon/routes/secret/details.js:24]: nit: personally I find this a bit hard to read. but could be pre-coffee brain
```suggestion
    const query = { backend, path };
    if (params.version) {
      query.version = params.version;
    }
```

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hashishaw) [reply ui/lib/kv/addon/routes/secret/metadata.js:63]: No, the capabilities will still return normally even if some of them require control group approval. But since this is only for metadata maybe we can catch it and fail silently? Most concerned about control groups for secret data access, which we aren't showing here

File: ui/lib/kv/addon/templates/secret/metadata/index.hbs
Comment (hashishaw) [inline ui/lib/kv/addon/templates/secret/metadata/index.hbs:10]: Again, not really sure why we need this capability since the action has already been taken

File: ui/lib/kv/addon/templates/secret/metadata/index.hbs
Comment (hashishaw) [inline ui/lib/kv/addon/templates/secret/metadata/index.hbs:16]: What if instead you passed in `@customMetadata`, and then you could eliminate handling the which case it is in the page component
```
@customMetadata={{or this.model.metadata.custom_metadata this.model.secret.custom_metadata}}
```

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js
Comment (hashishaw) [reply ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js:1450]: originURL is going to be the page they were on before they got to the page that requested the control group resource (so in this case, it's `/vault/secrets/${backend}/kv/${secretPathUrlEncoded}/paths`). Then the user needs to click the secret tab again and ensure that the secret data loads correctly.

**Review by hashishaw — COMMENTED** [review]
Nice work! A few comments, I think the only true blocker is the expected control group behavior

File: ui/app/adapters/kv/metadata.js
Comment (hellobontempo) [reply ui/app/adapters/kv/metadata.js:81]: I can rename! Doesn't matter to me 😄

File: ui/lib/kv/addon/components/page/list.js
Comment (hellobontempo) [reply ui/lib/kv/addon/components/page/list.js:90]: I think the diff may be confusing - this was the original workflow. It's when a user has permissions to read the secret but not list. What do you propose instead? The test was failing because navigating to the overview still works even though the secret isn't real. I don't think it makes sense to check whether or not the secret is real from the overview since we're trying to avoid make the request there.
Screenshot 2024-08-28 at 9 26 25 AM

https://github.com/hashicorp/vault/blob/main/ui/lib/kv/addon/components/page/list.js#L86C1-L91

File: ui/lib/kv/addon/components/page/secret/metadata/details.hbs
Comment (hellobontempo) [reply ui/lib/kv/addon/components/page/secret/metadata/details.hbs:51]: Sure - I was hesitant to change the workflow too much

File: ui/lib/kv/addon/routes/secret/details.js
Comment (hellobontempo) [reply ui/lib/kv/addon/routes/secret/details.js:24]: Optional spreading can be awkward to read! I'll update

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [reply ui/lib/kv/addon/routes/secret/metadata.js:63]: Yeah that makes sense! Thanks

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js
Comment (hellobontempo) [reply ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js:1450]: hmm, I had that originally and the test failed. 🤔  I'll try again

File: ui/lib/kv/addon/components/page/list.js
Comment (hashishaw) [reply ui/lib/kv/addon/components/page/list.js:90]: Right, I would expect it to go to the overview page, same as if you clicked on one of the list items. I agree it doesn't make sense to check if the secret exists. The screenshot you shared seems fine, and as a future improvement on that page we could show some error state if the metadata request comes back as 404

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hashishaw) [inline ui/lib/kv/addon/routes/secret/metadata.js:26]: Thinking about this more, maybe `fetchMultiplePaths` could return an object keyed by the paths, with the model as the value? just spitballing on the new pattern

File: ui/lib/kv/addon/templates/secret/metadata/index.hbs
Comment (hellobontempo) [reply ui/lib/kv/addon/templates/secret/metadata/index.hbs:10]: Previously we couldn't just rely on `this.customMetadata` since we didn't know if no metadata meant we can't read it or it doesn't exist. Now that the permissions have moved into the route and the `@metadata` or `@secret` don't exist or return `null` (before `secret` always existed and metadata returned an empty object) we can solely rely on the getter `this.customMetadata`

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js
Comment (hellobontempo) [reply ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-navigation-test.js:1450]: I think in my original attempt I had forgotten the `waitUntil`!

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [reply ui/lib/kv/addon/routes/secret/metadata.js:23]: Yes, it is!

File: ui/lib/kv/addon/routes/secret/metadata.js
Comment (hellobontempo) [reply ui/lib/kv/addon/routes/secret/metadata.js:26]: Yeah, that's a good idea. I was also wondering about just returning the permissions instead of the full model 🤔 If someone needs the list of capabilities, they can use the base `request` method

File: ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-create-test.js
Comment (hellobontempo) [inline ui/tests/acceptance/secrets/backend/kv/kv-v2-workflow-create-test.js:1011]: I deleted this test because this user's policy is:
```
path "secret/data/*" {
   capabilities = ["create", "update"]
}
path "secret/metadata/*" {
   capabilities = ["delete"]
}
```
with no read capabilities users cannot navigate to older versions of a secrert. if a user were to type in the version in the URL then this would result in a save error unless they happened to get the `cas` version right. 

instead I moved the warning assertion up to the create new version test above.

File: ui/app/services/capabilities.ts
Comment (hellobontempo) [inline ui/app/services/capabilities.ts:50]: Went the extra step of destructuring here so that just the `can` capabilities return. 
```js
// sample return
{
 "secret/data/my-secret": {
    canCreate: true,
    canDelete: false,
    canList: true,
    canPatch: true,
    canRead: true,
    canSudo: false,
    canUpdate: true,
 };
}
```

If we want to return the full capabilities model at some point (instead of an object with the capabilities) that's an easy refactor because the consumption of this method won't change. Retrieving the capability would still be `"secret/data/my-secret".canRead` etc.

File: ui/app/services/capabilities.ts
Comment (hashishaw) [inline ui/app/services/capabilities.ts:13]: 👏🏻

File: ui/lib/kv/addon/components/page/secret/metadata/details.hbs
Comment (hashishaw) [inline ui/lib/kv/addon/components/page/secret/metadata/details.hbs:51]: Great comment 👏🏻

**Review by hashishaw — APPROVED** [review]
Very nice! 🚀

**Pull request merged**
